File: backend/src/main/java/es/udc/tfgproject/backend/model/services/UserService.java
package es.udc.tfgproject.backend.model.services;

import es.udc.tfgproject.backend.model.entities.FavouriteAddress;
import es.udc.tfgproject.backend.model.entities.User;
import es.udc.tfgproject.backend.model.exceptions.DuplicateInstanceException;
import es.udc.tfgproject.backend.model.exceptions.IncorrectLoginException;
import es.udc.tfgproject.backend.model.exceptions.IncorrectPasswordException;
import es.udc.tfgproject.backend.model.exceptions.InstanceNotFoundException;
import es.udc.tfgproject.backend.model.exceptions.PermissionException;

public interface UserService {

	void signUp(User user) throws DuplicateInstanceException;

	void signUpBusinessman(User user) throws DuplicateInstanceException;

	User login(String userName, String password) throws IncorrectLoginException;

	User loginFromId(Long id) throws InstanceNotFoundException;

	User updateProfile(Long id, String firstName, String lastName, String email, String phone)
			throws InstanceNotFoundException;

	void changePassword(Long id, String oldPassword, String newPassword)
			throws InstanceNotFoundException, IncorrectPasswordException;

	FavouriteAddress addFavouriteAddress(String street, String cp, Long cityId, Long userId)
			throws InstanceNotFoundException;

	void deleteFavouriteAddress(Long userId, Long addressId) throws InstanceNotFoundException, PermissionException;

	Block<FavouriteAddress> findFavouriteAddresses(Long userId, int page, int size);

	Block<FavouriteAddress> findFavouriteAddressesByCity(Long userId, Long cityId, int page, int size);

	FavouriteAddress findFavAddress(Long userId, Long addressId) throws InstanceNotFoundException, PermissionException;

	void payWeeklyBalanceToCompanies(Long userId) throws InstanceNotFoundException, PermissionException;

	User payQuarterlyFee(Long userId) throws InstanceNotFoundException;
}

File: frontend/src/backend/userService.js
import {config, appFetch, setServiceToken, getServiceToken, removeServiceToken, setReauthenticationCallback} from './appFetch';

export const login = (userName, password, onSuccess, onErrors, reauthenticationCallback) =>
    appFetch('/users/login', config('POST', {userName, password}),
        authenticatedUser => {
            setServiceToken(authenticatedUser.serviceToken);
            setReauthenticationCallback(reauthenticationCallback);
            onSuccess(authenticatedUser);
        }, 
        onErrors);

export const tryLoginFromServiceToken = (onSuccess, reauthenticationCallback) => {

    const serviceToken = getServiceToken();

    if (!serviceToken) {
        onSuccess();
        return;
    }

    setReauthenticationCallback(reauthenticationCallback);

    appFetch('/users/loginFromServiceToken', config('POST'),
        authenticatedUser => onSuccess(authenticatedUser),
        () => removeServiceToken()
    );

}

export const signUp = (user, onSuccess, onErrors, reauthenticationCallback) => {

    appFetch('/users/signUp', config('POST', user), 
        authenticatedUser => {
            setServiceToken(authenticatedUser.serviceToken);
            setReauthenticationCallback(reauthenticationCallback);
            onSuccess(authenticatedUser);
        }, 
        onErrors);

}

export const signUpBusinessman = (user, onSuccess, onErrors, reauthenticationCallback) => {

    appFetch('/users/signUpBusinessman', config('POST', user), 
        authenticatedUser => {
            setServiceToken(authenticatedUser.serviceToken);
            setReauthenticationCallback(reauthenticationCallback);
            onSuccess(authenticatedUser);
        }, 
        onErrors);

}

export const logout = () => removeServiceToken();

export const updateProfile = (user, onSuccess, onErrors) =>
    appFetch(`/users/${user.id}`, config('PUT', user),
        onSuccess, onErrors);

export const changePassword = (id, oldPassword, newPassword, onSuccess,
    onErrors) =>
    appFetch(`/users/${id}/changePassword`, 
        config('POST', {oldPassword, newPassword}),
        onSuccess, onErrors);

export const addFavouriteAddress = (street, cp, cityId, onSuccess, onErrors) =>
appFetch(`/users/favouriteAddresses`,
    config('POST', {street, cp, cityId}),
    onSuccess, onErrors);

export const deleteFavouriteAddress = (addressId, onSuccess, onErrors) =>
    appFetch(`/users/favouriteAddresses/${addressId}`,
        config('DELETE'), onSuccess, onErrors);

export const findFavouriteAddresses = ({page}, onSuccess) => 
    appFetch(`/users/favouriteAddresses?page=${page}`, 
        config('GET'), onSuccess);

export const findFavAddress = (addressId, onSuccess) => 
    appFetch(`/users/favouriteAddress/${addressId}`, 
        config('GET'), onSuccess);

export const findFavouriteAddressesByCity = (cityId, {page}, onSuccess) => 
    appFetch(`/users/favouriteAddressesByCity?cityId=${cityId}&page=${page}`, 
        config('GET'), onSuccess);